File: optimization/src/test/scala/makina/scala/optimization/RosenbrockFunction.scala
package makina.scala.optimization

import breeze.linalg.{DenseMatrix, DenseVector}

/**
  * @author Emmanouil Antonios Platanios
  */
case class RosenbrockFunction() extends Function {
  override protected def computeValue(point: DenseVector[Double]): Double =
    100 * Math.pow(point(1) - Math.pow(point(0), 2), 2) + Math.pow(1 - point(0), 2)

  override protected def computeGradient(point: DenseVector[Double]): DenseVector[Double] =
    DenseVector(-400 * (point(1) - Math.pow(point(0), 2)) * point(0) - 2 * (1 - point(0)),
                200 * (point(1) - Math.pow(point(0), 2)))

  override protected def computeHessian(point: DenseVector[Double]): DenseMatrix[Double] =
    DenseMatrix((1200 * Math.pow(point(0), 2) - 400 * point(1) + 2, -400 * point(0)),
                (-400 * point(0), 200.0))
}

File: optimization/src/test/java/makina/optimization/RosenbrockFunction.java
package makina.optimization;

import makina.math.matrix.Matrix;
import makina.math.matrix.Vector;
import makina.math.matrix.Vectors;
import makina.optimization.function.AbstractFunction;

/**
 * @author Emmanouil Antonios Platanios
 */
class RosenbrockFunction extends AbstractFunction {
    @Override
    public double computeValue(Vector point) {
        double x1 = point.get(0);
        double x2 = point.get(1);
        return 100 * Math.pow(x2 - Math.pow(x1, 2), 2) + Math.pow(1 - x1, 2);
    }

    @Override
    public Vector computeGradient(Vector point) {
        double x1 = point.get(0);
        double x2 = point.get(1);
        double dx1 = - 400 * (x2 - Math.pow(x1, 2)) * x1 - 2 * (1 - x1);
        double dx2 = 200 * (x2 - Math.pow(x1, 2));
        return Vectors.dense(dx1, dx2);
    }

    @Override
    public Matrix computeHessian(Vector point) {
        double x1 = point.get(0);
        double x2 = point.get(1);
        double dx1x1 = 1200 * Math.pow(x1, 2) - 400 * x2 + 2;
        double dx1x2 = - 400 * x1;
        double dx2x2 = 200;
        return new Matrix(new double[][] { { dx1x1, dx1x2 }, { dx1x2, dx2x2 } });
    }
}
